File: cryptoPAN.cpp
// Implementação em C++ do algoritmo cripto-PAN

#include<vector>
#include <sstream>
#include "cryptoPAN.h"
using namespace std;
using namespace CryptoPP;

CryptoPanError::CryptoPanError(const std::string& msg) : message(msg) {}

const char* CryptoPanError::what() const noexcept {
    return message.c_str();
}

uint32_t CryptoPan::toInt(const std::array<uint8_t, 4>& arr) {
    return (arr[0] << 24) | (arr[1] << 16) | (arr[2] << 8) | arr[3];
}

CryptoPan::CryptoPan(const std::string& key) {

    // case of error
    if (key.size() != 32) {
        throw CryptoPanError("Key must be a 32 byte long string");
    }

    // Configurar AES com os primeiros 16 bytes da chave
    CryptoPP::SecByteBlock aesKey(reinterpret_cast<const uint8_t*>(key.data()), 16);
    aes.SetKey(aesKey, aesKey.size());


    // Gerar o padding usando os últimos 16 bytes da chave
    std::memcpy(pad.data(), key.data() + 16, 16);

    // Guardar processado
    aes.ProcessData(pad.data(), pad.data(),16);

    // Obter os primeiros 4 bytes de "pad" e convertê-los em inteiro
    std::array<uint8_t, 4> f4;
    std::copy(pad.begin(), pad.begin() + 4, f4.begin());

    uint32_t f4bp = toInt(f4);

    // Gerar máscaras e complementos
    for (int p = 0; p < 32; ++p) {
        uint32_t mask = (p == 0) ? 0 : (0xFFFFFFFF >> (32 - p)) << (32 - p);
        uint32_t complement = f4bp & (~mask);
        masks.emplace_back(mask, complement);
    }

}

array<uint8_t, 4> CryptoPan::toArray(uint32_t n) {
    return {static_cast<uint8_t>((n >> 24) & 0xFF),
            static_cast<uint8_t>((n >> 16) & 0xFF),
            static_cast<uint8_t>((n >> 8) & 0xFF),
            static_cast<uint8_t>(n & 0xFF)};
}

uint8_t CryptoPan::calc(uint32_t a) {
        array<uint8_t, 4> a_array = toArray(a);

        vector<uint8_t> inp(a_array.begin(), a_array.end());

        inp.insert(inp.end(), pad.begin() + 4, pad.end());

        vector<uint8_t> aes_output(CryptoPP::AES::BLOCKSIZE);
        aes.ProcessData(aes_output.data(), inp.data(), inp.size());

        uint8_t out = aes_output[0];

        return (out >> 7) & 0x01;
}


std::string CryptoPan::anonymize(const string& key, const string& ip) {
    vector<uint8_t> address;
    stringstream ss(ip);
    string segment;

    while (getline(ss, segment, '.')) {
        address.push_back(static_cast<uint8_t>(stoi(segment))); 
    }

    if (address.size() != 4) {
        throw CryptoPanError("Invalid IPv4 Address");
    }

    uint32_t newAddress = toInt({address[0], address[1], address[2], address[3]});

    vector<uint32_t> addresses;
    for (const auto& mask : masks) {
        uint32_t masked = (newAddress & mask.first) | mask.second;
        addresses.push_back(masked);
    }
   
    uint32_t result = 0;
    for (const auto& masked : addresses) {
        result = (result << 1) | calc(masked);
    }
    
    result = result ^ newAddress;
    
    array<uint8_t, 4> finalArray = toArray(result);


    stringstream ipStream;
    for (size_t i = 0; i < finalArray.size(); ++i) {
        if (i > 0) ipStream << ".";  
        ipStream << static_cast<int>(finalArray[i]);
    }

    return ipStream.str();

}

File: cryptopan.py
#!/usr/bin/env ipython3

# Versão modificada do código presente em https://github.com/FFM/pycryptopan/blob/master/cryptopan.py
# Esse software é livre e pode ser modificado, sem fins lucrativos, conforme a licença presente no repositório original

from functools import reduce
# from pycrypto module
from Crypto.Cipher.AES import new as AES
# from Crypto import Random


class CryptoPanError(Exception):
    def __init__(self, value):
        self.value=value

    def __str__(self):
        return repr(self.value)


class CryptoPan():
    def __init__(self, key):
        if len(key)!=32:
            raise CryptoPanError("Key must be a 32 byte long string")
        self.aes=AES(key[0:16],1)
        self.pad=self.aes.encrypt(key[16:32])

        f4=self.pad[0:4]

        # Python 2 requires explicit conversion to ints
        if isinstance(f4, str):
            f4=[ord(x) for x in f4]
        
        f4bp=self.toint(f4)
        self.masks=[(mask, f4bp & (~ mask)) for mask in (0xFFFFFFFF >> (32 - p) << (32 - p) for p in range(0, 32))]

    def toint(self, array):
        return array[0] << 24 | array[1] <<16 | array[2] << 8 | array[3]

    def toarray(self, n):
        for i in range(3, -1, -1):
            yield (n >> (i * 8)) & 0xFF

    def anonymize(self, ip):
        result=0
        address=[int(x) for x in ip.split(".")]
        if len(address)!=4:
            raise CryptoPanError("Invalid IPv4 Address")

        address=self.toint(address)

        def calc(a):
            """ calculate the first bit for Crypto-PAN"""
            a_array = self.toarray(a)
                
            # Python 2 requires converting ints to chars one at a time
            if isinstance(self.pad, str):
                inp="".join((chr(x) for x in a_array))
            else:
                inp=bytes(a_array)

            inp+=self.pad[4:]
            
            rin_output=self.aes.encrypt(inp)
            out=rin_output[0]

            # Python 2 requires explicit conversion to int
            if isinstance(out, str):
                out=ord(out)

            return out >> 7
        
        addresses = [(address & mask[0]) | mask[1] for mask in self.masks]
        result=reduce(lambda x, y: x << 1 | y, (calc(a) for a in addresses), 0)
        return ".".join(["%s" % x for x in self.toarray(result ^ address)])
    

def main():
    key = b"boojahyoo3vaeToong0Eijee7Ahz3yee"
    crypto_pan = CryptoPan(key)
    anonymized_ip = crypto_pan.anonymize("192.0.2.1")
    print(anonymized_ip)
    
    # a resposta deve ser "206.2.124.120" conforme o teste da biblioteca original


if __name__ == "__main__":
    main()